### DISCUSSION TITLE
New import resolution

### DISCUSSION THREAD

**Discussion by lerno** [c3lang/c3c#455] [General]
In `dev` imports work like this:

1. `import foo::bar;` will import `foo::bar` and any sub packages of `foo::bar`
2. For any package, it will without import be able to access any packages with the same top package name. So `foo::bar` can access `foo::baz::my_function()` without an import.

Try it out. Currently you're even required to import std::builtin. This will probably change.

**Comment by data-man** [comment]
```rust
import libc;

fn void main()
{
    libc::printf("Hi!\n");
}
```
`$ c3c compile test.c3`
`$ ls -1 *.o`
Why are so many `*.o` files created?
>libc.errno.o
libc.o
math.o
std.array.o
std.atomic.o
std.builtin.o
std.cinterop.o
std.env.o
std.hash.adler32.o
std.hash.crc32.o
std.hash.crc64.o
std.io.o
std.math.o
std.mem.o
std.os.linux.o
std.os.macos.o
std.os.windows.o
std.runtime.o
std.str.o
test.o

**Reply by lerno** [reply]
Because currently the whole standard library is compiled as well. There are possible optimizations, some which are incompatible

(1) precompile the standard library once as a static library, add it at link time
(2) cache standard library .o files
(3) only compile the part of the standard library that is in use
(4) only codegen the parts of the standard library that is in use

**Comment by data-man** [comment]
Deleting temporary files after compilation is annoying. My vote is for gcc/clang behavior.
And `clean` command doesn't work as expected.

**Reply by lerno** [reply]
> Deleting temporary files after compilation is annoying. My vote is for gcc/clang behavior.
> And `clean` command doesn't work as expected.

I added an issue for that. #456
